File: py/app/converters/anthropic_to_gemini.py
"""Convert Anthropic format to Gemini format."""

import json
from typing import Any

from py.app.models import anthropic as ant
from py.app.models import gemini as gem


def convert_request(req: ant.MessagesRequest) -> tuple[gem.GenerateContentRequest, str]:
    """Convert Anthropic messages request to Gemini generate content request.

    Returns (request, model_name) tuple.
    """
    # System instruction
    system_instruction = None
    if req.system:
        system_instruction = gem.SystemInstruction(
            parts=[gem.TextPart(text=req.system)]
        )

    # Convert messages
    contents: list[gem.Content] = []

    for msg in req.messages:
        if msg.role == "user":
            if isinstance(msg.content, str):
                contents.append(
                    gem.Content(role="user", parts=[gem.TextPart(text=msg.content)])
                )
            else:
                parts: list[gem.Part] = []
                for block in msg.content:
                    if isinstance(block, ant.TextContent):
                        parts.append(gem.TextPart(text=block.text))
                    elif isinstance(block, ant.ToolResultContent):
                        content = (
                            block.content
                            if isinstance(block.content, str)
                            else block.content[0].text
                        )
                        # Need to find the tool name - we'll use a placeholder
                        parts.append(
                            gem.FunctionResponsePart(
                                functionResponse={
                                    "name": "",  # Will need context to fill this
                                    "response": json.loads(content)
                                    if content.startswith("{")
                                    else {"result": content},
                                }
                            )
                        )
                if parts:
                    contents.append(gem.Content(role="user", parts=parts))

        elif msg.role == "assistant":
            if isinstance(msg.content, str):
                contents.append(
                    gem.Content(role="model", parts=[gem.TextPart(text=msg.content)])
                )
            else:
                parts = []
                for block in msg.content:
                    if isinstance(block, ant.TextContent):
                        parts.append(gem.TextPart(text=block.text))
                    elif isinstance(block, ant.ToolUseContent):
                        parts.append(
                            gem.FunctionCallPart(
                                functionCall={
                                    "name": block.name,
                                    "args": block.input,
                                }
                            )
                        )
                if parts:
                    contents.append(gem.Content(role="model", parts=parts))

    # Convert tools
    tools = None
    if req.tools:
        function_declarations = []
        for tool in req.tools:
            function_declarations.append(
                gem.FunctionDeclaration(
                    name=tool.name,
                    description=tool.description,
                    parameters={
                        "type": tool.input_schema.type,
                        "properties": tool.input_schema.properties,
                        "required": tool.input_schema.required,
                    },
                )
            )
        tools = [gem.ToolConfig(functionDeclarations=function_declarations)]

    # Generation config
    generation_config = gem.GenerationConfig(
        temperature=req.temperature,
        topP=req.top_p,
        topK=req.top_k,
        maxOutputTokens=req.max_tokens,
        stopSequences=req.stop_sequences,
    )

    return (
        gem.GenerateContentRequest(
            contents=contents,
            systemInstruction=system_instruction,
            generationConfig=generation_config,
            tools=tools,
        ),
        req.model,
    )


def convert_response(resp: gem.GenerateContentResponse, model: str) -> ant.MessagesResponse:
    """Convert Gemini generate content response to Anthropic messages response."""
    content: list[ant.TextContent | ant.ToolUseContent] = []

    if resp.candidates:
        candidate = resp.candidates[0]
        for i, part in enumerate(candidate.content.parts):
            if isinstance(part, gem.TextPart):
                content.append(ant.TextContent(text=part.text))
            elif isinstance(part, gem.FunctionCallPart):
                fc = part.functionCall
                content.append(
                    ant.ToolUseContent(
                        id=f"toolu_{i}",
                        name=fc.get("name", ""),
                        input=fc.get("args", {}),
                    )
                )

    # Map finish reason
    finish_reason = resp.candidates[0].finishReason if resp.candidates else "STOP"
    stop_reason_map = {
        "STOP": "end_turn",
        "MAX_TOKENS": "max_tokens",
        "SAFETY": "end_turn",
        "RECITATION": "end_turn",
    }
    stop_reason = stop_reason_map.get(finish_reason or "STOP", "end_turn")

    # Check for tool use
    has_tool_use = any(isinstance(c, ant.ToolUseContent) for c in content)
    if has_tool_use:
        stop_reason = "tool_use"

    return ant.MessagesResponse(
        id=f"msg_gemini_{id(resp)}",
        model=model,
        content=content if content else [ant.TextContent(text="")],
        stop_reason=stop_reason,
        usage=ant.ResponseUsage(
            input_tokens=resp.usageMetadata.promptTokenCount if resp.usageMetadata else 0,
            output_tokens=resp.usageMetadata.candidatesTokenCount if resp.usageMetadata else 0,
        ),
    )


def convert_stream_chunk(data: dict[str, Any], msg_id: str) -> str | None:
    """Convert Gemini stream chunk to Anthropic stream event.

    Returns SSE formatted string or None if chunk should be skipped.
    """
    candidates = data.get("candidates", [])
    if not candidates:
        return None

    candidate = candidates[0]
    content = candidate.get("content", {})
    parts = content.get("parts", [])

    result = ""

    for part in parts:
        if "text" in part:
            event = {
                "type": "content_block_delta",
                "index": 0,
                "delta": {"type": "text_delta", "text": part["text"]},
            }
            result += f"event: content_block_delta\ndata: {json.dumps(event)}\n\n"

    finish_reason = candidate.get("finishReason")
    if finish_reason:
        stop_reason_map = {
            "STOP": "end_turn",
            "MAX_TOKENS": "max_tokens",
            "SAFETY": "end_turn",
        }
        stop_reason = stop_reason_map.get(finish_reason, "end_turn")
        event = {
            "type": "message_delta",
            "delta": {"stop_reason": stop_reason},
            "usage": {"output_tokens": 0},
        }
        result += f"event: message_delta\ndata: {json.dumps(event)}\n\n"
        result += "event: message_stop\ndata: {\"type\": \"message_stop\"}\n\n"

    return result if result else None

File: internal/converters/anthropic_to_gemini.go
package converters

import (
	"encoding/json"

	"ai_gateway/internal/models"
)

// AnthropicToGeminiRequest converts an Anthropic request to Gemini format
func AnthropicToGeminiRequest(req *models.MessagesRequest) (*models.GenerateContentRequest, error) {
	geminiReq := &models.GenerateContentRequest{}

	// Set generation config
	geminiReq.GenerationConfig = &models.GenerationConfig{}
	if req.Temperature != nil {
		geminiReq.GenerationConfig.Temperature = req.Temperature
	}
	if req.TopP != nil {
		geminiReq.GenerationConfig.TopP = req.TopP
	}
	if req.TopK != nil {
		geminiReq.GenerationConfig.TopK = req.TopK
	}
	if req.MaxTokens > 0 {
		geminiReq.GenerationConfig.MaxOutputTokens = &req.MaxTokens
	}
	if len(req.StopSequences) > 0 {
		geminiReq.GenerationConfig.StopSequences = req.StopSequences
	}

	// Convert system message
	if req.System != nil {
		var systemText string
		switch v := req.System.(type) {
		case string:
			systemText = v
		case []interface{}:
			for _, block := range v {
				if blockMap, ok := block.(map[string]interface{}); ok {
					if blockMap["type"] == "text" {
						systemText += getString(blockMap, "text")
					}
				}
			}
		}
		if systemText != "" {
			geminiReq.SystemInstruction = &models.GeminiContent{
				Parts: []models.GeminiPart{{Text: systemText}},
			}
		}
	}

	// Convert messages
	var contents []models.GeminiContent
	for _, msg := range req.Messages {
		geminiContent := models.GeminiContent{}

		// Map role
		switch msg.Role {
		case "assistant":
			geminiContent.Role = "model"
		default:
			geminiContent.Role = "user"
		}

		// Handle content
		switch content := msg.Content.(type) {
		case string:
			geminiContent.Parts = []models.GeminiPart{{Text: content}}
		case []interface{}:
			for _, block := range content {
				if blockMap, ok := block.(map[string]interface{}); ok {
					blockType := getString(blockMap, "type")
					switch blockType {
					case "text":
						geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{
							Text: getString(blockMap, "text"),
						})
					case "tool_use":
						args, _ := blockMap["input"].(map[string]interface{})
						geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{
							FunctionCall: &models.GeminiFunctionCall{
								Name: getString(blockMap, "name"),
								Args: args,
							},
						})
					case "tool_result":
						var responseContent interface{}
						if c, ok := blockMap["content"].(string); ok {
							json.Unmarshal([]byte(c), &responseContent)
						} else {
							responseContent = blockMap["content"]
						}
						// Tool results go in a user message
						geminiContent.Role = "user"
						geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{
							FunctionResponse: &models.FunctionResponse{
								Name:     getString(blockMap, "name"),
								Response: map[string]interface{}{"result": responseContent},
							},
						})
					case "image":
						if source, ok := blockMap["source"].(map[string]interface{}); ok {
							geminiContent.Parts = append(geminiContent.Parts, models.GeminiPart{
								InlineData: &models.InlineData{
									MimeType: getString(source, "media_type"),
									Data:     getString(source, "data"),
								},
							})
						}
					}
				}
			}
		}

		if len(geminiContent.Parts) > 0 {
			contents = append(contents, geminiContent)
		}
	}
	geminiReq.Contents = contents

	// Convert tools
	if len(req.Tools) > 0 {
		var declarations []models.FunctionDeclaration
		for _, tool := range req.Tools {
			declarations = append(declarations, models.FunctionDeclaration{
				Name:        tool.Name,
				Description: tool.Description,
				Parameters:  tool.InputSchema,
			})
		}
		geminiReq.Tools = []models.GeminiTool{{
			FunctionDeclarations: declarations,
		}}
	}

	return geminiReq, nil
}

// GeminiToAnthropicResponse converts a Gemini response to Anthropic format
func GeminiToAnthropicResponse(resp map[string]interface{}, model string) (*models.MessagesResponse, error) {
	anthropicResp := &models.MessagesResponse{
		ID:    generateID(),
		Type:  "message",
		Role:  "assistant",
		Model: model,
	}

	candidates, ok := resp["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return anthropicResp, nil
	}

	candidate := candidates[0].(map[string]interface{})
	content, ok := candidate["content"].(map[string]interface{})
	if !ok {
		return anthropicResp, nil
	}

	parts, ok := content["parts"].([]interface{})
	if !ok {
		return anthropicResp, nil
	}

	var contentBlocks []models.ContentBlock
	for _, part := range parts {
		partMap := part.(map[string]interface{})
		if text, ok := partMap["text"].(string); ok {
			contentBlocks = append(contentBlocks, models.ContentBlock{
				Type: "text",
				Text: text,
			})
		}
		if fc, ok := partMap["functionCall"].(map[string]interface{}); ok {
			contentBlocks = append(contentBlocks, models.ContentBlock{
				Type:  "tool_use",
				ID:    generateToolCallID(len(contentBlocks)),
				Name:  getString(fc, "name"),
				Input: fc["args"],
			})
		}
	}
	anthropicResp.Content = contentBlocks

	// Convert finish reason
	if fr, ok := candidate["finishReason"].(string); ok {
		var stopReason string
		switch fr {
		case "STOP":
			stopReason = "end_turn"
		case "MAX_TOKENS":
			stopReason = "max_tokens"
		default:
			if len(contentBlocks) > 0 {
				for _, block := range contentBlocks {
					if block.Type == "tool_use" {
						stopReason = "tool_use"
						break
					}
				}
			}
			if stopReason == "" {
				stopReason = "end_turn"
			}
		}
		anthropicResp.StopReason = &stopReason
	}

	// Convert usage
	if usage, ok := resp["usageMetadata"].(map[string]interface{}); ok {
		anthropicResp.Usage = models.AnthropicUsage{
			InputTokens:  getInt(usage, "promptTokenCount"),
			OutputTokens: getInt(usage, "candidatesTokenCount"),
		}
	}

	return anthropicResp, nil
}

// GeminiStreamToAnthropicStream converts a Gemini stream event to Anthropic format
func GeminiStreamToAnthropicStream(data map[string]interface{}, isFirst bool, model string) ([][]byte, error) {
	var events [][]byte

	candidates, ok := data["candidates"].([]interface{})
	if !ok || len(candidates) == 0 {
		return nil, nil
	}

	candidate := candidates[0].(map[string]interface{})
	content, ok := candidate["content"].(map[string]interface{})
	if !ok {
		return nil, nil
	}

	parts, ok := content["parts"].([]interface{})
	if !ok || len(parts) == 0 {
		return nil, nil
	}

	if isFirst {
		// Send message_start event
		startEvent := map[string]interface{}{
			"type": "message_start",
			"message": map[string]interface{}{
				"id":          generateID(),
				"type":        "message",
				"role":        "assistant",
				"content":     []interface{}{},
				"model":       model,
				"stop_reason": nil,
				"usage":       map[string]interface{}{"input_tokens": 0, "output_tokens": 0},
			},
		}
		startBytes, _ := json.Marshal(startEvent)
		events = append(events, startBytes)

		// Send content_block_start
		blockStartEvent := map[string]interface{}{
			"type":  "content_block_start",
			"index": 0,
			"content_block": map[string]interface{}{
				"type": "text",
				"text": "",
			},
		}
		blockStartBytes, _ := json.Marshal(blockStartEvent)
		events = append(events, blockStartBytes)
	}

	part := parts[0].(map[string]interface{})
	if text, ok := part["text"].(string); ok {
		deltaEvent := map[string]interface{}{
			"type":  "content_block_delta",
			"index": 0,
			"delta": map[string]interface{}{
				"type": "text_delta",
				"text": text,
			},
		}
		deltaBytes, _ := json.Marshal(deltaEvent)
		events = append(events, deltaBytes)
	}

	// Handle finish
	if fr, ok := candidate["finishReason"].(string); ok && fr != "" {
		// content_block_stop
		stopBlockEvent := map[string]interface{}{
			"type":  "content_block_stop",
			"index": 0,
		}
		stopBlockBytes, _ := json.Marshal(stopBlockEvent)
		events = append(events, stopBlockBytes)

		// message_delta
		var stopReason string
		switch fr {
		case "STOP":
			stopReason = "end_turn"
		case "MAX_TOKENS":
			stopReason = "max_tokens"
		default:
			stopReason = "end_turn"
		}

		messageDeltaEvent := map[string]interface{}{
			"type": "message_delta",
			"delta": map[string]interface{}{
				"stop_reason": stopReason,
			},
			"usage": map[string]interface{}{
				"output_tokens": 0,
			},
		}
		messageDeltaBytes, _ := json.Marshal(messageDeltaEvent)
		events = append(events, messageDeltaBytes)

		// message_stop
		messageStopEvent := map[string]interface{}{
			"type": "message_stop",
		}
		messageStopBytes, _ := json.Marshal(messageStopEvent)
		events = append(events, messageStopBytes)
	}

	return events, nil
}
